sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteRename.java

```diff
@@ -58,51 +60,65 @@ public WriteRename(
       PCollectionView<String> jobIdToken,
       WriteDisposition writeDisposition,
       CreateDisposition createDisposition,
+      @Nullable
       PCollectionView<Map<TableDestination, Iterable<String>>> tempTablesView) {
     this.bqServices = bqServices;
     this.jobIdToken = jobIdToken;
-    this.writeDisposition = writeDisposition;
-    this.createDisposition = createDisposition;
+    this.firstPaneWriteDisposition = writeDisposition;
+    this.firstPaneCreateDisposition = createDisposition;
     this.tempTablesView = tempTablesView;
   }
 
   @ProcessElement
   public void processElement(ProcessContext c) throws Exception {
-    Map<TableDestination, Iterable<String>> tempTablesMap =
-        Maps.newHashMap(c.sideInput(tempTablesView));
-
-    // Process each destination table.
-    for (Map.Entry<TableDestination, Iterable<String>> entry : tempTablesMap.entrySet()) {
-      TableDestination finalTableDestination = entry.getKey();
-      List<String> tempTablesJson = Lists.newArrayList(entry.getValue());
-      // Do not copy if no temp tables are provided
-      if (tempTablesJson.size() == 0) {
-        return;
-      }
+    if (tempTablesView != null) {
+      Map<TableDestination, Iterable<String>> tempTablesMap =
+          Maps.newHashMap(c.sideInput(tempTablesView));
 
-      List<TableReference> tempTables = Lists.newArrayList();
-      for (String table : tempTablesJson) {
-        tempTables.add(BigQueryHelpers.fromJsonString(table, TableReference.class));
+      // Process each destination table.
+      for (Map.Entry<TableDestination, Iterable<String>> entry : tempTablesMap.entrySet()) {
+        writeRename(entry.getKey(), entry.getValue(), c);
       }
+    } else {
+      writeRename(c.element().getKey(), c.element().getValue(), c);
+    }
+  }
+
+  private void writeRename(TableDestination finalTableDestination,
+                           Iterable<String> tempTableNames,
+                           ProcessContext c) throws Exception {
+    WriteDisposition writeDisposition =
+        (c.pane().getIndex() == 0) ? firstPaneWriteDisposition : WriteDisposition.WRITE_APPEND;
```

Comment:
This is essentially the definition of pane index. It's the count of how many panes have been processed prior to this.